File: day3.R
# Decription of script ---------------------------------------------------------
# Day 3 advent of code -- Toboggan Trajectory
#
# ------------------------------------------------------------------------------

# set up -----------------------------------------------------------------------
library(tidyverse)

# load data --------------------------------------------------------------------

geology <- read_table(paste0(here::here(), "/day3.txt"), col_names = FALSE)

geology_transformed <- geology %>%
  separate(X1, paste("col", 1:nchar(geology$X1[1])), sep = "(?<=.)", extra = "drop")

# solve the puzzle -------------------------------------------------------------
# Part 1:- count the number of trees encountered if you take the path
# 3 across and 1 down

nrows <- nrow(geology_transformed)
ncols <- ncol(geology_transformed)
route_col <- 3*(1:nrows)
relative_postition <- route_col%%ncols + 1

tree_count <- 0
for (i in 2:nrows){
  tree_count <- tree_count + (geology_transformed[i, relative_postition[i-1]] == "#")
}

# Part 2
# now try different combinations of slopes
# Right 1, down 1.
# Right 3, down 1. (This is the slope you already checked.)
# Right 5, down 1.
# Right 7, down 1.
# Right 1, down 2.

# create a function to work out the position of the slope
slope_function <- function(right, down){
  route_col <- right*(1:round(nrows/down))
  relative_postition <- route_col%%ncols + 1
  tree_count <- 0
  for (i in 1:(round(nrows/down)-1)){
    tree_count <- tree_count + (geology_transformed[1 + (down * i), relative_postition[i]] == "#")
  }
  return(tree_count)
}

# multiple the slopes together
slope_function(1,1) * slope_function(3,1) * slope_function(5,1) * slope_function(7,1) * slope_function(1,2)

File: day3.py
# Description ------------------------------------------------------------------
# Advent of code day 2 -- find the valid passwords based on different policies 
# Attempted in R first and then converted into python. For this one I used a better 
# solution that I found online in R to convert to python. 

# set up -----------------------------------------------------------------------

# load data --------------------------------------------------------------------
def split(word): 
    return [char for char in word.rstrip('\n')]  

with open('day3.txt') as file:
    path = [split(text) for text in file]

# solve the puzzle -------------------------------------------------------------
# Part 1 count the number of trees encountered if you take the path
# 3 across and 1 down

route_row = [1 + x for x in range(len(path)-1)]
route_col = [(3 * x) % len(path[1]) for x in range(1,len(path))]

number_of_trees = 0
for i in route_row:
    number_of_trees += path[i][route_col[i-1]] == '#'

# Part 2 # now try different combinations of slopes
# Right 1, down 1.
# Right 3, down 1. (This is the slope you already checked.)
# Right 5, down 1.
# Right 7, down 1.
# Right 1, down 2.

def traverse(right, down):
    route_row = [1 + x for x in range(round(len(path)/down)-1)]
    route_col = [(right * x) % len(path[1]) for x in range(1,round(len(path)/down))]

    number_of_trees = 0
    for i in route_row:
        number_of_trees += path[i*down][route_col[i-1]] == '#'
    return number_of_trees

traverse(1,1) * traverse(3,1) * traverse(5,1) * traverse(7,1) * traverse(1,2)

